Improve Crossgen2 performance via sorting improvements

davidwrighton opened the pull request:
- Crossgen2 makes extensive use of sorting to achieve determinism
- Replace most signifcant sort operations with a parallel mergesort
  - These sorts use rather expensive comparison functions and as such the standard sort algorithm which is tuned more to reduce memory usage than reduce comparisons is not ideal, and merge sort works better
  - Primarily sorting is performed during the previously single threaded portions of the product. Adding parallelism yields easy wins
  - In a large benchmark overall compilation performance improved from 106 seconds to 82 seconds
- Implementation makes extensive use of generics to share the implementation across the various apis that are used for sorting in crossgen2
- Add Detection on of duplicates in sort algorithm
- Perf result compiling a large set of assemblies
  - Using standard sorts 121.6 seconds
  - Using parallel merge sort 96.9 seconds

davidwrighton committed:
Improve Crossgen2 performance
- Crossgen2 makes extensive use of sorting to achieve determinism
- Replace most signifcant sort operations with a parallel mergesort
  - These sorts use rather expensive comparison functions and as such the standard sort algorithm which is tuned more to reduce memory usage than reduce comparisons is not ideal, and merge sort works better
  - Primarily sorting is performed during the previously single threaded portions of the product. Adding parallelism yields easy wins
  - In a large benchmark overall compilation performance improved from 106 seconds to 82 seconds
- Implementation makes extensive use of generics to share the implementation across the various apis that are used for sorting in crossgen2
- Add Detection on of duplicates in sort algorithm
- Perf result compiling a large set of assemblies
  - Using standard sorts 121.6 seconds
  - Using parallel merge sort 96.9 seconds

davidwrighton commented:
@dotnet/crossgen-contrib

nattress commented:
Do you have any idea of the working set regression you're seeing in a large compile - say framework composite?

trylek commented on src/coreclr/src/tools/Common/Sorting/MergeSort.cs:
```diff
@@ -0,0 +1,185 @@
+        }
+
+        // Parallel sorting api which will sort in parallel when appropriate
+        public static void ParallelSort<T>(this List<T> listToSort, Comparison<T> comparison)
```
Nit - could we just use ISortableDataStructureAccessor as the "listToSort" to avoid having to duplicate the method so many times?

trylek commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ArrayOfEmbeddedDataNode.cs:
```diff
@@ -88,7 +88,7 @@ public override ObjectData GetData(NodeFactory factory, bool relocsOnly)
 
             if (_sorter != null)
-                _nestedNodesList.Sort(_sorter);
+                _nestedNodesList.ParallelSort(_sorter);
```
Hmm, wouldn't calling it <code>.MergeSort</code> be a tad more descriptive given how you call it in the implementation?

trylek reviewed (approved):
Looks great, thanks David for driving this! I guess that, as a next step, I would be interested why our comparers are so slow; in the few cases I recall off the top of my head we should be doing structured comparisons, not something as stupid as blindly emitting a textual representation of a node and subsequently using plain lexicographic sorting...

davidwrighton commented:
@nattress The extra temporary allocations from this algorithm are about the same size as the initial data structure to sort. In the context of crossgen2, which the graph is often quite large, the sortable structure itself is a small fraction of the size, and the size cost is really small. I didn't see a memory size regression of significance, although the test cases you've been looking at are much bigger, and may show a more significant regression.

@trylek The reason the comparers are slow is that the objects they are comparing are spread all over the GC heap, with extremely poor locality. Its not really a matter of performing any individual comparison faster, its more about reducing the latency in reading the interesting bits of memory.

davidwrighton committed:
Address code review feedback
- Rename to MergeSort
- Refactor to remove code duplication
- Drop the support for falling back to standard sort algorithms

MichalStrehovsky commented:
> Using standard sorts 121.6 seconds
> Using parallel merge sort 96.9 seconds

This is a pretty surprising bit of information. I won't have time to look into this, but as a datapoint - the CoreRT AOT compiler is architected pretty similarly (compilation is multithreaded and the sorting of everything happens on a single thread).

In the CoreRT compiler the sorting contributes to only ~2% of wallclock time (I did a quick measurement for "CoreRT compiler compiling CoreRT compiler" because that's what I was playing with last weekend - it's a bigger app than a Hello world, but of course smaller than many others).

I'm kind of wondering why ReadyToRun needs to sort so much more and whether the main problem is that we're generating too many things.

jakobbotsch commented on src/coreclr/src/tools/Common/Sorting/MergeSortCore.cs:
```diff
@@ -0,0 +1,130 @@
+                T[] localCopyOfHalfOfArray = new T[halfLen];
+                accessor.Copy(arrayToSort, index, localCopyOfHalfOfArray, 0, halfLen);
+                await MergeSortCore<T, T[], ArrayAccessor<T>, TComparer, TCompareAsEqualAction>.ParallelSort(localCopyOfHalfOfArray, 0, halfLen, comparer);
+                await rightSortComplete.Task;
```
Why not save the task returned from `Task.Run` and await that here?

davidwrighton replied:
The Task returned by `Task.Run` completes at the first await, and not when the SetResult completes, therefore it isn't useful to await that.

jakobbotsch replied:
The `Func<Task>` overload of `Task.Run` returns a proxy task that represents the full completion ([MSDN](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.run?view=netcore-3.1#System_Threading_Tasks_Task_Run__1_System_Func_System_Threading_Tasks_Task___0___)).

davidwrighton commented:
@MichalStrehovsky yes, it is clear that the expansion rate of R2R images is greater than it should be, and that is the real problem. I'm looking into that as well, as it leads to more extreme problems like oversize files, and such. However, this fix is nice as it substantially reduces the cycle time of looking at changes. I suspect that what is happening is that crossgen2 is getting TLB cache thrashing, which can be disastrous for performance, and this fix is spreading that load across various different cores thus leading to significantly better perf.

vectorix commented:
This method doesn't look right.

         public void SwapElements(T[] dataStructure, int i, int i2)
        {
            T temp = dataStructure[i];
            dataStructure[i] = dataStructure[i + 1];
            dataStructure[i + 1] = temp;
        }

It doesn't swap i with i2, it swaps i with i +1 and ignores i2.

Since the caller passes i + 1 in i2 this parameter may be redundant, but then the name should probably be something else, e.g. SwapWithNext.

davidwrighton commented:
@vectorix wow, that's an interesting copy/paste bug. Fortunately, the only use of this function is to swap with next, so the behavior is … correct in the product, but a function that wrong should be fixed. Thank you for your review.
